src/Features/Core/Portable/ExtractClass/AbstractExtractClassRefactoringProvider.cs

```diff
@@ -0,0 +1,107 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable 
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.PullMemberUp;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.ExtractClass
+{
+    internal abstract class AbstractExtractClassRefactoringProvider : CodeRefactoringProvider
+    {
+        public AbstractExtractClassRefactoringProvider()
+        {
+        }
+
+        public AbstractExtractClassRefactoringProvider(IExtractClassOptionsService service)
+        {
+            _optionsService = service;
+        }
+
+        private readonly IExtractClassOptionsService? _optionsService;
+
+        protected abstract Task<SyntaxNode?> GetSelectedNodeAsync(CodeRefactoringContext context);
+        protected abstract Task<SyntaxNode?> GetSelectedClassDeclarationAsync(CodeRefactoringContext context);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var optionsService = _optionsService ?? context.Document.Project.Solution.Workspace.Services.GetService<IExtractClassOptionsService>();
+            if (optionsService is null)
+            {
+                return;
+            }
+
+            // If we register the action on a class node, no need to find selected members. Just allow
+            // the action to be invoked with the dialog 
+            if (await TryRegisterClassActionAsync(context, optionsService).ConfigureAwait(false))
+            {
+                return;
+            }
+
+            await RegisterMemberActionAsync(context, optionsService).ConfigureAwait(false);
```

Comment:
consider making less imperative.

```
var action = await TryGetClassActionAsync(...) ?? await TryGetMemberActionAsync(...);
if (action != null)
    RegisterAction
```

Reply:
this doesnt' actually seem to ahve been resolved.  it's stily very imperative.

Reply:
Changed to get actions instead of registering them